vehicle-python/src/vehicle_lang/ast/__init__.py

```diff
@@ -59,373 +59,459 @@ class Tensor(Generic[DType]):
+    value: Any
+
+    def __init__(self) -> None:
+        raise TypeError("Cannot instantiate abstract class")
+
+
+@dataclass(frozen=True, init=False)
+class BuiltinConstant(AST):
+    # Abstract base for constant symbols
+    def __init__(self) -> None:
+        raise TypeError("Cannot instantiate abstract class")
+
+
+@dataclass(frozen=True, init=False)
+class BuiltinType(AST):
+    # Abstract base for built-in type symbols
     def __init__(self) -> None:
-        raise TypeError("Cannot instantiate abstract class BuiltinFunction")
+        raise TypeError("Cannot instantiate abstract class")
+
+
+@dataclass(frozen=True, init=False)
+class Expression(AST):
+    # Abstract base for expression nodes
+    def __init__(self) -> None:
+        raise TypeError("Cannot instantiate abstract class")
+
+
+@dataclass(frozen=True, init=False)
+class BuiltinFunction(Expression):
+    # Abstract base for built-in function symbols
+    def __init__(self) -> None:
+        raise TypeError("Cannot instantiate abstract class")
+
+
+@dataclass(frozen=True, init=False)
+class Declaration(AST, metaclass=ABCMeta):
+    # Abstract base for top-level declarations
+    def __init__(self) -> None:
+        raise TypeError("Cannot instantiate abstract class")
+
+    def get_name(self) -> Name:
+        # Abstract method to get the name of the declaration
+        raise NotImplementedError("Subclasses must implement get_name()")
+
+
+# Type Alias for common Union used in function arguments/bodies
+# ==============================================================================
+
+FunctionInput: TypeAlias = Union[
+    BuiltinFunction, Expression, BuiltinConstant, BuiltinLiteral
+]
+
+
+################################################################################
+# Builtin Literals
+################################################################################
 
 
 @dataclass(frozen=True)
-class ConsList(BuiltinFunction):
-    pass
+class Index(BuiltinLiteral):
+    value: int
 
 
 @dataclass(frozen=True)
-class NotBoolTensor(BuiltinFunction):
-    pass
+class BoolTensor(BuiltinLiteral):
+    value: Tensor[bool]
 
 
 @dataclass(frozen=True)
-class AndBoolTensor(BuiltinFunction):
-    pass
+class NatTensor(BuiltinLiteral):
+    value: Tensor[int]
 
 
 @dataclass(frozen=True)
-class OrBoolTensor(BuiltinFunction):
-    pass
+class IntTensor(BuiltinLiteral):
+    value: Tensor[int]
 
 
 @dataclass(frozen=True)
-class NegRatTensor(BuiltinFunction):
-    pass
+class RatTensor(BuiltinLiteral):
+    value: Tensor[Fraction]
 
 
 @dataclass(frozen=True)
-class AddRatTensor(BuiltinFunction):
-    pass
+class RatLiteral(BuiltinLiteral):
+    value: Fraction
+
+
+################################################################################
+# Builtin Constants
+################################################################################
 
 
 @dataclass(frozen=True)
-class SubRatTensor(BuiltinFunction):
+class Unit(BuiltinConstant):
     pass
```

Comment:
This is the AST we're trying to mimic: https://github.com/vehicle-lang/vehicle/blob/dev/vehicle/src/Vehicle/Backend/LossFunction/JSON.hs

I don't think we need a `Unit` class now?